### PR TITLE
[RNMobile] add animation to FloatingToolbar

### PR DESCRIPTION
**Pull request by jbinda**
## Description

After change FloatingToolbar position we decide to add some entry/exit animation according to [comment with animation guideline](https://github.com/WordPress/gutenberg/pull/21446#issuecomment-615417890)

Please also refer to:
[Related gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2225)

## How has this been tested?

1. Add below code in `initial-html`

<details>

```
/**
 * @format
 * @flow
 */

export default `
<!-- wp:media-text {"mediaId":1,"mediaType":"image"} -->
<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="https://cldup.com/cXyG__fTLN.jpg" class="wp-image-1"/></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"className":"has-large-font-size"} -->
<p class="has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->

<!-- wp:image -->
<figure class="wp-block-image"><img src="https://cldup.com/cXyG__fTLN.jpg" alt=""/></figure>
<!-- /wp:image -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:image {"id":1,"sizeSlug":"large"} -->
<figure class="wp-block-image size-large"><img src="https://cldup.com/cXyG__fTLN.jpg" alt="" class="wp-image-1"/></figure>
<!-- /wp:image -->

<!-- wp:heading -->
<h2>Test heading</h2>
<!-- /wp:heading --></div></div>
<!-- /wp:group -->
`;
```

</details>

2. Feel free to add more nested structure
3. Check if each time you select from root list block to nested block FloatingToolbar is pop ups with ease in animation on bottom(iOS) / top (Android) of the screen
4. Check if each time you select from nested block to root list block FloatingToolbar is disappears with ease out animation

## Types of changes

New feature - animation on FloatingToolbar

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jbinda** [comment]
@iamthomasbishop

I need to wait on jitpack fix to generate production build on Android.
Production build on iOS is here: [IPA 28171](https://github.com/wordpress-mobile/WordPress-iOS/pull/14070#issuecomment-625246976)

File: packages/block-editor/src/components/floating-toolbar/index.native.js [hunk 1]
```diff
@@ -18,30 +25,88 @@ import styles from './styles.scss';
...
+	const [ previousSelection, setPreviousSelection ] = useState( {} );
+
+	useEffect( () => {
+		const easing = Easing.ease;
+		Animated.timing( opacity, {
+			toValue: showFloatingToolbar ? 1 : 0,
+			duration: showFloatingToolbar
+				? EASE_IN_DURATION
+				: EASE_OUT_DURATION,
+			easing,
```
Comment (dratwas) [inline packages/block-editor/src/components/floating-toolbar/index.native.js:51]: I see no point in creating the easing const to use it here.

```suggestion
			Easing.ease,
```

File: packages/block-editor/src/components/floating-toolbar/index.native.js [hunk 2]
```diff
@@ -18,30 +25,88 @@ import styles from './styles.scss';
...
-				) }
-				<Breadcrumb clientId={ selectedClientId } />
-			</View>
-		</TouchableWithoutFeedback>
+} ) => {
+	// sustain old selection for proper Breacdrumb rendering when exit animation is ongoing
+	const [ previousSelection, setPreviousSelection ] = useState( {} );
+
+	useEffect( () => {
+		const easing = Easing.ease;
```
Comment (dratwas) [inline packages/block-editor/src/components/floating-toolbar/index.native.js:45]:
```suggestion
```

File: packages/block-editor/src/components/floating-toolbar/index.native.js [hunk 3]
```diff
@@ -18,30 +25,88 @@ import styles from './styles.scss';
...
+		Animated.timing( opacity, {
```
Comment (dratwas) [inline packages/block-editor/src/components/floating-toolbar/index.native.js:46]: Please add `useNativeDriver: true` here :)

https://reactnative.dev/docs/animations#using-the-native-driver

Comment (jbinda) [reply packages/block-editor/src/components/floating-toolbar/index.native.js:45]: done

File: packages/block-editor/src/components/floating-toolbar/index.native.js (hunk 1 above)
Comment (jbinda) [reply packages/block-editor/src/components/floating-toolbar/index.native.js]: done

File: packages/block-editor/src/components/floating-toolbar/index.native.js (hunk 2 above)
Comment (jbinda) [reply packages/block-editor/src/components/floating-toolbar/index.native.js]: done

File: packages/block-editor/src/components/floating-toolbar/index.native.js [hunk 4]
```diff
@@ -18,30 +25,88 @@ import styles from './styles.scss';
...
+	const showPrevious = previousSelectedClientId && ! showFloatingToolbar;
+	const breadcrumbClientId = showPrevious
+		? previousSelectedClientId
+		: selectedClientId;
+	const showNavUpButton =
+		!! parentId || ( showPrevious && !! previousSelectedParentId );
+
+	return (
+		!! opacity && (
+			<TouchableWithoutFeedback accessible={ false }>
```
Comment (dratwas) [inline packages/block-editor/src/components/floating-toolbar/index.native.js:87]: Just out of curiosity, do we really need that `TouchableWithoutFeedback`? What would happen if we remove it and leave the `Animated.View` as a wrapper?

Comment (jbinda) [reply packages/block-editor/src/components/floating-toolbar/index.native.js:87]: I think it make sense. One thing to check if a11y will work as expected after. Will try to cover that today

Comment (jbinda) [reply packages/block-editor/src/components/floating-toolbar/index.native.js:87]: I have tested the a11y on both platform with that solution and it works fine so I will apply changes

Comment (dratwas) [reply packages/block-editor/src/components/floating-toolbar/index.native.js]: Awesome, thanks :)

**Review by dratwas — APPROVED** [review]

**Comment by designsimply** [comment]
Noticed that the enter/exit FloatingToolbar animations work as described except when going from one group block to another. ([42s](https://cloudup.com/c7xNBkPeZVr)) I think this is the expected behavior and am just making a note of it for reference. 👍

**Comment by iamthomasbishop** [comment]
Nice catch! You're right, this is intended behavior, but I agree with your assessment. It might be worth trying a cross-fading transition between the two. My original thought while working on this was that the transition between the two might be jarring if it very quickly exits and re-enters — that it could look like a glitch. But I think cross-fading the contents of the toolbar (icons, labels) might act as a nice subtle improvement.

@dratwas @lukewalczak perhaps we can iterate on this? Obviously it's not a top priority, so no major rush.

**Comment by designsimply** [comment]
My note wasn't a push for a change btw! It's more of a note-to-self so if this comes up in the future, I'll know why the animation wasn't there the first time I tried testing it (because I happened to switch from one group block to another). 🙂 I like the transition either way and it didn't feel too jarring to me because the label shape changed with the label name when I switched from one group to another group block.

**Comment by iamthomasbishop** [comment]
> My note wasn't a push for a change btw!

@designsimply Oh, I know, but it was a good suggestion! 😄

**Pull request merged**